golem/README.md

```diff
@@ -0,0 +1,308 @@
+
+3. **sbt** installed
+
+4. A reachable Golem router/executor (local or cloud, depending on your `GOLEM_CLI_FLAGS`)
+
+### Define Your Agent
+
+Create a trait describing your agent's interface:
+
+```scala
+import cloud.golem.runtime.annotations.{DurabilityMode, agentDefinition, description}
+import cloud.golem.sdk.BaseAgent
+import zio.blocks.schema.Schema
+
+import scala.concurrent.Future
+
+// Define your data types
+final case class Name(value: String)
+
+object Name {
+  implicit val schema: Schema[Name] = Schema.derived
+}
+
+// Define your agent trait (typeName is optional; when omitted, it is derived from the trait name)
+@agentDefinition(mode = DurabilityMode.Durable)
+@description("A simple name-processing agent")
+trait NameAgent extends BaseAgent {
+  type AgentInput = Unit // or a case class with a Schema
+
+  @description("Reverse the provided name")
+  def reverse(input: Name): Future[Name]
+}
+```
+
+### Implement and Register
+
+```scala
+import cloud.golem.runtime.autowire._
+
+object NameAgentModule {
+  private class NameAgentImpl extends NameAgent {
+    override def reverse(input: Name): Name =
+      input.copy(value = input.value.reverse)
+  }
+
+  // Type name is derived from @agentDefinition(...) on the trait:
+  val definition: AgentDefinition[NameAgent] =
+    AgentImplementation.register[NameAgent](new NameAgentImpl)
+}
+```
+
+### Connect as a Client
+
+From another component, connect to a remote agent:
+
+```scala
+import cloud.golem.runtime.rpc.AgentClient
+import scala.concurrent.Future
+
+val plan = AgentClient.plan[NameAgent] // uses @agentDefinition + NameAgent.AgentInput
+
+// Connect and invoke
+val result: Future[NameAgent] = AgentClient.connect(plan, ())
+```
+
+### Optional companion ergonomics (Scala-only)
+
+If you want `Shard.get(...)` / `Shard.getPhantom(...)` style ergonomics, Scala requires a companion `object Shard` to exist.
+Today this is a one-liner:
+
+```scala
+import cloud.golem.runtime.annotations.{DurabilityMode, agentDefinition, agentImplementation, description}
+import cloud.golem.runtime.autowire.{AgentDefinition, AgentImplementation}
+import cloud.golem.sdk.{AgentCompanion, BaseAgent, Uuid}
+
+import scala.concurrent.Future
+
+@agentDefinition(mode = DurabilityMode.Durable)
+trait Shard extends BaseAgent {
+  type AgentInput = (String, Int)
+
+  @description("Get a value from the table")
+  def get(key: String): Future[Option[String]]
+
+  @description("Set a value in the table")
+  def set(key: String, value: String): Unit
+}
+
+object Shard extends AgentCompanion[Shard]
+
+@agentImplementation()
+final class ShardImpl(input: (String, Int)) extends Shard {
+  override def get(key: String): Future[Option[String]] = Future.successful(None)
+  override def set(key: String, value: String): Unit = ()
+}
+
+object ShardModule {
+  val definition: AgentDefinition[Shard] =
+    AgentImplementation.register[Shard, (String, Int)](in => new ShardImpl(in))
```

Comment:
I hope user never need to manually register the agent implementation in the code. :)